pkg/mcs/resourcemanager/server/manager.go

```diff
@@ -406,155 +427,52 @@ func (m *Manager) backgroundMetricsFlush(ctx context.Context) {
-			}
-			// CPU time info.
-			if consumption.TotalCpuTimeMs > 0 {
-				if consumption.SqlLayerCpuTimeMs > 0 {
-					sqlLayerRuMetrics.Add(consumption.SqlLayerCpuTimeMs * m.controllerConfig.RequestUnit.CPUMsCost)
-					sqlCPUMetrics.Add(consumption.SqlLayerCpuTimeMs)
-				}
-				kvCPUMetrics.Add(consumption.TotalCpuTimeMs - consumption.SqlLayerCpuTimeMs)
-			}
-			// RPC count info.
-			if consumption.KvReadRpcCount > 0 {
-				readRequestCountMetrics.Add(consumption.KvReadRpcCount)
-			}
-			if consumption.KvWriteRpcCount > 0 {
-				writeRequestCountMetrics.Add(consumption.KvWriteRpcCount)
-			}
-
-			m.consumptionRecord[consumptionRecordKey{keyspaceID: keyspaceID, name: name, ruType: ruLabelType}] = time.Now()
+			name := consumptionInfo.resourceGroupName
+			getMaxPerSecTracker(keyspaceID, keyspaceName, name).collect(consumption)
+			getCounterMetrics(keyspaceID, keyspaceName, name, ruLabelType).add(consumption, m.controllerConfig)
+			insertConsumptionRecord(keyspaceID, name, ruLabelType)
 
 			// TODO: maybe we need to distinguish background ru.
 			if rg := m.GetMutableResourceGroup(keyspaceID, name); rg != nil {
 				rg.UpdateRUConsumption(consumptionInfo.Consumption)
 			}
 		case <-cleanUpTicker.C:
 			// Clean up the metrics that have not been updated for a long time.
-			for r, lastTime := range m.consumptionRecord {
+			for r, lastTime := range consumptionRecordMap {
 				if time.Since(lastTime) <= metricsCleanupTimeout {
 					continue
 				}
-				readRequestUnitCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				writeRequestUnitCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				sqlLayerRequestUnitCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				readByteCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				writeByteCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				kvCPUCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				sqlCPUCost.DeleteLabelValues(r.name, r.name, r.ruType)
-				requestCount.DeleteLabelValues(r.name, r.name, readTypeLabel)
-				requestCount.DeleteLabelValues(r.name, r.name, writeTypeLabel)
-				availableRUCounter.DeleteLabelValues(r.name, r.name)
-				delete(m.consumptionRecord, r)
-				deleteMaxPerSecTracker(r.keyspaceID, r.name)
-				readRequestUnitMaxPerSecCost.DeleteLabelValues(r.name)
-				writeRequestUnitMaxPerSecCost.DeleteLabelValues(r.name)
-				resourceGroupConfigGauge.DeletePartialMatch(prometheus.Labels{newResourceGroupNameLabel: r.name})
+				keyspaceName, err := m.getKeyspaceNameByID(ctx, r.keyspaceID)
+				if err != nil {
+					continue
+				}
+				deleteConsumptionRecord(r)
+				deleteMetrics(r.keyspaceID, keyspaceName, r.groupName, r.ruType)
+				deleteMaxPerSecTracker(r.keyspaceID, r.groupName)
 			}
 		case <-availableRUTicker.C:
 			// Prevent from holding the lock too long when there're many keyspaces and resource groups.
 			for _, krgm := range m.getKeyspaceResourceGroupManagers() {
+				keyspaceName, err := m.getKeyspaceNameByID(ctx, krgm.keyspaceID)
```

Comment:
Is it possible to store the keyspace name in the manager?

Reply:
You mean keyspace manager? I tried, but encountered a circular reference problem, and considering that the resource manager microservice might be deployed in the future, it seems that we cannot fully rely on calls that may exist in another process's logic.